ffi/diplomat/src/errors.rs

```diff
@@ -0,0 +1,147 @@
+        WriteableError = 1,
+
+        // Some input was out of bounds
+        OutOfBoundsError = 2,
+
+        // general data errors
+        // See DataError
+        DataMissingResourceKeyError = 10,
+        DataMissingVariantError = 11,
+        DataMissingLocaleError = 12,
+        DataMissingResourceOptionsError = 13,
+        DataNeedsVariantError = 14,
+        DataNeedsLocaleError = 15,
+        DataExtraneousResourceOptionsError = 16,
+        DataFilteredResourceError = 17,
+        DataMismatchedTypeError = 18,
+        DataMissingPayloadError = 19,
+        DataInvalidStateError = 20,
+        DataCustomError = 21,
+        DataIoError = 22,
+        DataUnavailableBufferFormatError = 23,
+
+        // locale and data struct errors
+        /// The subtag being requested was not set
+        LocaleUndefinedSubtagError = 31,
+        /// The locale or subtag string failed to parse
+        LocaleParserError = 32,
+        /// Attempted to construct an invalid data struct
+        DataStructValidityError = 33,
+
+        // property and decimal errors
+        PropertyUnknownScriptIdError = 40,
+        PropertyUnknownGeneralCategoryGroupError = 41,
+        DecimalLimitError = 42,
+        DecimalSyntaxError = 43,
+
+        // plural errors
+        PluralParserError = 50,
+    }
+}
+
+impl From<fmt::Error> for ICU4XError {
+    fn from(_: fmt::Error) -> Self {
+        ICU4XError::WriteableError
+    }
+}
+
+impl From<DataError> for ICU4XError {
+    fn from(e: DataError) -> Self {
+        match e.kind {
+            DataErrorKind::MissingResourceKey => ICU4XError::DataMissingResourceKeyError,
+            DataErrorKind::MissingVariant => ICU4XError::DataMissingVariantError,
+            DataErrorKind::MissingLocale => ICU4XError::DataMissingLocaleError,
+            DataErrorKind::MissingResourceOptions => ICU4XError::DataMissingResourceOptionsError,
+            DataErrorKind::NeedsVariant => ICU4XError::DataNeedsVariantError,
+            DataErrorKind::NeedsLocale => ICU4XError::DataNeedsLocaleError,
+            DataErrorKind::ExtraneousResourceOptions => {
+                ICU4XError::DataExtraneousResourceOptionsError
+            }
+            DataErrorKind::FilteredResource => ICU4XError::DataFilteredResourceError,
+            DataErrorKind::MismatchedType(..) => ICU4XError::DataMismatchedTypeError,
+            DataErrorKind::MissingPayload => ICU4XError::DataMissingPayloadError,
+            DataErrorKind::InvalidState => ICU4XError::DataInvalidStateError,
+            DataErrorKind::Custom => ICU4XError::DataCustomError,
+            #[cfg(all(
+                feature = "provider_fs",
+                not(any(target_arch = "wasm32", target_os = "none"))
+            ))]
+            DataErrorKind::Io(..) => ICU4XError::DataIoError,
+            // datagen only
+            // DataErrorKind::MissingSourceData(..) => ..,
+            DataErrorKind::UnavailableBufferFormat(..) => {
```

Comment:
Nit: For cases where you drop info, this is where I think we should `log` the output. Maybe we should do this in general in all of the `From<blah> for ICU4XError` impls?

Reply:
Hmm, might need to be a feature, but yeah.

Reply:
Having a way to enable logging was the compromise reached to get me on board with error enums. :)

Reply:
Yeah I recall! I hadn't planned to do that in this PR but could do it. Prefer to do it in a followup because then I can do it Monday :smile:

Reply:
So I started to look at it and I think it should be its own PR: if we're using the `log` crate we need to initialize the logger somehow, and it's unclear if we should just expose env_logger or do something else. WASM is fine, but everyone else is tricky.

So we'd need a bit of scaffolding for this.

Reply:
OK to do in a follow-up. That would be cleaner.